src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -72,96 +72,82 @@ namespace ts.refactor.convertToOptionalChainExpression {
+        const finalExpression = getFinalExpressionInChain(expression.right);
 
-        if (!lastPropertyAccessChain) return undefined;
+        if (!finalExpression) return undefined;
 
-        const occurrences = getOccurrencesInExpression(lastPropertyAccessChain.expression, expression.left, checker);
-        return occurrences ? { lastPropertyAccessChain, occurrences, expression } : undefined;
+        const occurrences = getOccurrencesInExpression(finalExpression.expression, expression.left);
+        return occurrences ? { finalExpression, occurrences, expression } : undefined;
     }
 
     /**
      * Gets a list of property accesses that appear in matchTo and occur in sequence in expression.
      */
-    function getOccurrencesInExpression(matchTo: Expression, expression: Expression, checker: TypeChecker): (PropertyAccessExpression | Identifier)[] | undefined {
+    function getOccurrencesInExpression(matchTo: Expression, expression: Expression): (PropertyAccessExpression | Identifier)[] | undefined {
         const occurrences: (PropertyAccessExpression | Identifier)[] = [];
         while (isBinaryExpression(expression) && expression.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken) {
-            const match = getMatchingSubexpression(matchTo, expression.right, checker);
+            const match = getMatchingStart(matchTo, expression.right);
             if (!match) {
                 break;
             }
             occurrences.push(match);
             matchTo = match;
             expression = expression.left;
         }
-        const finalMatch = getMatchingSubexpression(matchTo, expression, checker);
+        const finalMatch = getMatchingStart(matchTo, expression);
         if (finalMatch) {
             occurrences.push(finalMatch);
         }
         return occurrences.length > 0 ? occurrences: undefined;
     }
 
     /**
-     * Checks that expression is a syntactic subexpression of matchTo.
+     * Returns subchain if chain begins with subchain syntactically.
      */
-    function getMatchingSubexpression(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
-        if (isCallExpression(matchTo)) {
-            return !isCallExpression(expression) ? getMatchingSubexpression(matchTo.expression, expression, checker) : undefined;
-        }
-        else if ((isPropertyAccessExpression(expression) || isIdentifier(expression)) && (isPropertyAccessExpression(matchTo) || isIdentifier(matchTo))) {
-            return containsSyntacticSubchain(matchTo, expression, checker) ? expression : undefined;
-        }
-        return undefined;
+    function getMatchingStart(chain: Expression, subchain: Expression): PropertyAccessExpression | Identifier | undefined {
+        if (!isIdentifier(subchain) && !isPropertyAccessExpression(subchain)) return undefined;
+        return chainStartsWith(chain, subchain) ? subchain : undefined;

```

Comment:
Would this work as well? Seems easier to read than double `!`

```suggestion
        return (isIdentifier(subchain) || isPropertyAccessExpression(subchain)) && chainStartsWith(chain, subchain) ? subchain : undefined;
```